doc/developer/design/20240609_error_handling.md

```diff
@@ -0,0 +1,302 @@
+   errors, or eliminate errors in future releases of Materialize.
+
+   Note also that `ONLY ERRORS` doesn't fundamentally add new backwards
+   compatibility surface area. Error messages, descriptions, hints, and codes
+   are already returned via the SQL interface, and users may already be
+   erroneously relying on seeing particular error messages and codes from
+   Materialize in particular situations.
+
+### Implementation
+
+#### `IGNORE ERRORS`
+
+[@maddyblue] has a prototype implementation of `IGNORE ERRORS` in [#27137].
+The core of the implementation is the addition of a new `ignore_errors`
+field to `MirRelationExpr::Get`:
+
+```diff
+ MirRelationExpr::Get {
+     /// A global or local identifier naming the collection.
+     id: Id,
+
+     // ...
+
++    /// Whether to ignore errors.
++    ignore_errors: bool,
+ },
+```
+
+When rendered, if `ignore_errors` is true, the referenced collection's errs
+stream is simply not wired up to the downstream error stream.
+
+**Open questions:**
+
+  * Is this implementation acceptable to the cluster team?
+  * Will the existence of the new field cause correctness issues if existing
+    references to `MirRelationExpr` fail to check the field?
+
+#### `ONLY ERRORS`
+
+There is not yet a prototype for `ONLY ERRORS`.
+
+One candidate proposal is to introduce a `FlavoredGid` type, which couples a
+global ID with a specification of whether the oks stream or the errs stream is
+of interest.
+
+```diff
++enum GidFlavor {
++    Oks,
++    Errs,
++}
++
++struct FlavoredGid {
++    id: GlobalId,
++    flavor: GidFlavor,
++}
++
+ pub enum Id {
+     Local(LocalId),
+-    Global(GlobalId),
++    Global(FlavoredGid),
+ }
+```
+
+This would be an unfortunately massive change. We'd need to audit every use of
+`GlobalId` across the codebase to assess whether it should remain a `GlobalId`
+or become a `FlavoredGid`.
+
+## Alternatives
+
+### Kafka source-specific dead letter queue
+
+Only the Kafka source is particularly prone to bad data. With PostgreSQL and
+MySQL sources, it is almost always a bug (either in in the upstream system or in
+Materialize) if data fails to decode, as the upstream systems enforce schemas.
+But Kafka does not guarantee schema enforcement, so it is common for bad data to
+slip into a Kafka topic.
+
+Because Kafka sources are the most prone to bad data, we could instead choose to
+pursue a Kafka source-specific solution. For example, imagine that we introduced
+a dead letter queue (DLQ) for Kafka sources. Concretely, we could add a `DEAD
+LETTERS` option (strawman name) to `CREATE SOURCE` that specifies the name of a
+relation in which to emit undecodable messages:
+
+```
+CREATE SOURCE ksrc TO KAFKA CONNECTION kconn ... WITH (
+    DEAD LETTERS = kbad
+)
+```
+
+The `kbad` relation would have a structure like:
+
+ Name         | Type
+--------------|--------
+ `key`        | `bytea`
+ `value`      | `bytea`
+ `error`      | `text`
+ `error_code` | `text`
```

Comment:
It occurs to me that one benefit of the DLQ approach is that we could add a `ingested_at` column here that describes when the error occurred. Not sure we have that same luxury with the `ONLY ERRORS` approach ... unless we never advanced the since frontier of that collection?

In general there's a tangle of considerations here around "how do I ignore errors in the DLQ that I've already seen?"

Also need to think through how errors would get _retracted_ from this collection, if at all, for upsert sources.